aptos-move/move-examples/marketplace/sources/token_offer.move

```diff
@@ -0,0 +1,725 @@
+    public entry fun sell_tokenv2<CoinType>(
+        seller: &signer,
+        token_offer: Object<TokenOffer>,
+        token: Object<TokenV2>,
+    ) acquires CoinOffer, TokenOffer, TokenOfferTokenV1, TokenOfferTokenV2 {
+        let token_offer_addr = token_offer_addr<TokenOfferTokenV2>(&token_offer);
+
+        // Move the token to its destination
+
+        let recipient = object::owner(token_offer);
+        object::transfer(seller, token, recipient);
+
+        // Pay fees
+
+        let royalty = tokenv2::royalty(token);
+        let (royalty_payee, royalty_denominator, royalty_numerator) = if (option::is_some(&royalty)) {
+            let royalty = option::destroy_some(royalty);
+            let payee_address = royalty::payee_address(&royalty);
+            let denominator = royalty::denominator(&royalty);
+            let numerator = royalty::numerator(&royalty);
+            (payee_address, denominator, numerator)
+        } else {
+            (signer::address_of(seller), 1, 0)
+        };
+
+        settle_payments<CoinType>(
+            signer::address_of(seller),
+            token_offer_addr,
+            royalty_payee,
+            royalty_denominator,
+            royalty_numerator,
+        );
+    }
+
+    /// From the coin offer remove appropriate payment for the token and distribute to the seller,
+    /// the creator for royalties, and the marketplace for commission. If there are no more slots,
+    /// cleanup the offer.
+    inline fun settle_payments<CoinType>(
+        seller: address,
+        token_offer_addr: address,
+        royalty_payee: address,
+        royalty_denominator: u64,
+        royalty_numerator: u64,
+    ) acquires CoinOffer, TokenOffer, TokenOfferTokenV1, TokenOfferTokenV2 {
+        assert!(exists<TokenOffer>(token_offer_addr), ENO_TOKEN_OFFER);
+        let token_offer_obj = borrow_global_mut<TokenOffer>(token_offer_addr);
+        assert!(
+            timestamp::now_seconds() < token_offer_obj.expiration_time,
+            error::invalid_state(EEXPIRED),
+        );
+        let price = token_offer_obj.item_price;
+
+        assert!(
+            exists<CoinOffer<CoinType>>(token_offer_addr),
+            error::not_found(ENO_COIN_OFFER),
+        );
+        let coin_offer = borrow_global_mut<CoinOffer<CoinType>>(token_offer_addr);
+        let coins = coin::extract(&mut coin_offer.coins, price);
+
+        let royalty_charge = price * royalty_numerator / royalty_denominator;
+        let royalties = coin::extract(&mut coins, royalty_charge);
+        coin::deposit(royalty_payee, royalties);
+
+        let fee_schedule = token_offer_obj.fee_schedule;
+        let commission_charge = fee_schedule::commission(fee_schedule, price);
+        let commission = coin::extract(&mut coins, commission_charge);
+        coin::deposit(fee_schedule::fee_address(fee_schedule), commission);
+
+        coin::deposit(seller, coins);
+
+        fee_schedule::emit_token_offer_event(
```

Comment:
We probably also want to emit events when token and collection offers are canceled.

Reply:
There's some data that needs to be indexed for deletes, like canceled price and token metadata.